src/user-safe/UserSafe.sol

```diff
@@ -0,0 +1,678 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {SafeERC20, IERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20Permit} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Permit.sol";
+import {ICashDataProvider} from "../interfaces/ICashDataProvider.sol";
+import {SignatureUtils} from "../libraries/SignatureUtils.sol";
+import {ISwapper} from "../interfaces/ISwapper.sol";
+import {IPriceProvider} from "../interfaces/IPriceProvider.sol";
+import {OwnableUpgradeable} from "openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import {Initializable} from "openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+
+/**
+ * @title UserSafe
+ * @author ether.fi [shivam@ether.fi]
+ * @notice User safe account for interactions with the EtherFi Cash contracts
+ */
+contract UserSafe is Initializable, OwnableUpgradeable {
+    using SafeERC20 for IERC20;
+    using SignatureUtils for bytes32;
+
+    enum SpendingLimitTypes {
+        Daily,
+        Weekly,
+        Monthly,
+        Yearly
+    }
+
+    struct WithdrawalRequest {
+        address[] tokens;
+        address recipient;
+        uint96 finalizeTime;
+    }
+
+    struct WithdrawalData {
+        address[] tokens;
+        uint256[] amounts;
+        address recipient;
+        uint96 finalizeTime;
+    }
+
+    struct SpendingLimitData {
+        SpendingLimitTypes spendingLimitType;
+        uint64 renewalTimestamp;
+        uint256 spendingLimit; // in USD with 6 decimals
+        uint256 usedUpAmount; // in USD with 6 decimals
+    }
+
+    bytes32 public constant REQUEST_WITHDRAWAL_METHOD =
+        keccak256("requestWithdrawal");
+    bytes32 public constant APPROVE_METHOD = keccak256("approve");
+    bytes32 public constant SET_SPENDING_LIMIT_METHOD =
+        keccak256("setSpendingLimit");
+    bytes32 public constant SET_INCOMING_SPENDING_LIMIT_METHOD =
+        keccak256("setIncomingSpendingLimit");
+
+    // Address of the USDC token
+    address public immutable usdc;
+    // Address of the weETH token
+    address public immutable weETH;
+    // Address of the Cash Data Provider
+    ICashDataProvider public immutable cashDataProvider;
+    // Address of the price provider
+    IPriceProvider public immutable priceProvider;
+    // Address of the swapper
+    ISwapper public immutable _swapper;
+    // Withdrawal requests pending with the contract
+    WithdrawalRequest private _pendingWithdrawalRequest;
```

Comment:
can there be only one pending withdrawal at a time? 

or if users request multiple, they are merged? then, how do you handle their different finalize time?

Reply:
but i like it, since it is simple than managing per-request queue

Reply:
I think its best that there is only one request which can be pending at any point in time.

Reply:
agreed. let's start simple and extend later if needed

Reply:
Sure

Reply:
resolved